### ISSUE TITLE
Add flexibility for .err and .out comparison in cabal-install integration tests

### ISSUE THREAD

**Issue by BardurArantsson** [haskell/cabal#2872]
As discussed in https://github.com/haskell/cabal/pull/2868#discussion_r42120492 it would be nice to have a little more flexibility for the matching used for ".err" and ".out" files in the cabal-install integration tests. This would allow for better handling of Windows (".exe" appears in the "program name" portion of output) and perhaps also in other cases where a *little* flexibility is required.

(Mostly copied from that thread:)

A reasonable start would be line-by-line regex comparison where each .out or .err file could contain lines like:

```
   RE:this is a regex
   this is not
```

If we want to go super-advanced, something like http://llvm.org/docs/CommandGuide/FileCheck.html could also be an option. (I think this may be overkill for our purposes since cabal-install output is *usually* pretty simple if you can control the inputs.)

**Comment by BardurArantsson** [comment]
@23Skidoo : Turns out, I can't assign myself. Oh, GitHub Issues how I love thee :).

**Comment by 23Skidoo** [comment]
@BardurArantsson Just gave you and @phadej Write permissions, so self-assigning should work from now on.

**Comment by 23Skidoo** [comment]
One thing I like about FileCheck is that both the test and the expected output are combined in a single file, so tests are more self-contained. In principle, we could do the same thing.

Also, if we want to reuse the same approach in the test suite for Cabal proper, then it probably makes sense to package `IntegrationTests.hs` as a separate library (say, `tasty-cabal-integration-tester`).

**Comment by BardurArantsson** [comment]
I don't think the "integration-test" approach works at all for Cabal itself, really. While doing the integration-tests thing I had a quick look at the PackageTests for Cabal, and it appears that the Cabal `PackageTester.hs` (and sometimes the `Check.hs` for each individual test) contains basically a more hardcoded subset of `cabal-install` which is necessary to run the checks.

I don't think we'll (realistically) ever have fully self-contained tests -- after all we'll always need .cabal files and directories with sources, etc. etc.

A bit off-topic, but seeing the redundancy between Cabal/PackageTests and cabal-install reminds me: Overall, it appears that the Cabal/cabal-install split is a very strange and not *entirely* helpful one. It seems to be split in a sort of horizontal way:

```
|   (world)     |
|---------------|
| cabal-install |
|---------------|
|    Cabal      |
|---------------|
```

whereas I think it's often more helpful to split things in a vertical way, like (hypothetically):

```
|  (world)     |  (world)   |   (world)   | ...
|--------------+------------+-------------+
|              |            |  commands   | ...
|--------------+------------+-------------+
| file-formats |   solver   |             | ...
|--------------+------------+-------------+
```

where `commands` could be sort of like the current `cabal-install` (perhaps as a library) and would depend on `solver`, `file-formats`, and whatever else it needed. The main point here would be to separate along *functionality* lines rather than some artificial "abstraction level".

Anyway, this should probably be on the mailing list and not here...

**Comment by ezyang** [comment]
@BardurArantsson The historic motivation for the abstraction barrier is that Cabal looks at one package at a time, cabal-install looks at multiple packages. One might argue that Cabal/cabal-install should be modularized further (the file-formats library), but historically it's been pretty annoying to do development on multiple packages so it's not been done.

**Comment by BardurArantsson** [comment]
@ezyang Oh, right. That does kind of make sense -- so cabal-install essentially started as a "solver" of sorts and evolved into the beast we have today :).

**Comment by BardurArantsson** [comment]
Just FTR: It's pretty unlikely that I'll actually get around to this any time in the near future (and the demand is probably quite limited), so if anyone wants to pick it up, be my guest. I'd be happy to help.

**Comment by BardurArantsson** [comment]
(Added label "easy" to the 'start-of-line-regex-marker' version since it requires very little knowledge of Cabal internals and would probably be sufficient for most uses.)

**Comment by ezyang** [comment]
@BardurArantsson I'm attempting to write a test with your new format, and I don't understand, how are you supposed to actually get the intended output? Is there an `make accept` option like in GHC? I don't see any code which would support that.

**Comment by BardurArantsson** [comment]
Not sure I understand the question... but assuming I do: you just add a ".out" and ".err" file with the same name as the test script and the test harness will automatically compare them against the expected output on stdout/stderr (respectively).

**Comment by ezyang** [comment]
I guess, Cabal generates a lot of output and it was not obvious how to actually copy paste that output into the file.

But based on looking at existing tests, it looks like the recommended method is to `/dev/null` or `-v0` the tests until the output is at a short and manageable level. I'm not sure how I feel about this; when the test goes wrong I usually WANT to see the output; this just guarantees I'm going to have to rerun the test manually.

**Comment by BardurArantsson** [comment]
Yeah, so one method I used was to just have an empty .err or .out file while developing the tests. That'll the full output of Cabal and you can decide from there what you want to keep and what you can throw away (with /dev/null redirect).

Because of cabal-install's verbosity it's pretty hard to make *everything* line up perfectly (esp. given the variable nature of some of the output). I guess one idea might be to actually have an environment variable which the "cabal" invocation in `common.sh` could look at to decide if it should show everything. Then it'd at least be easier to toggle the non-essential output on/off.

**Comment by BardurArantsson** [comment]
This was effectively closed by #3130, so closing.

(If there really is a need for even more flexible matching, it can be reopened.)
